.rubocop_todo.yml

```diff
@@ -1,26 +1,718 @@
+# AllowedMethods: instance_of?, kind_of?, is_a?, eql?, respond_to?, equal?
+Lint/RedundantSafeNavigation:
+  Exclude:
+    - 'lib/puppet/parser/functions/assert_private.rb'
+    - 'lib/puppet/parser/functions/getparam.rb'
+
+# Offense count: 4
+# Configuration parameters: AllowComments, AllowNil.
 Lint/SuppressedException:
   Exclude:
-    - 'lib/facter/facter_dot_d.rb'
     - 'lib/puppet/functions/merge.rb'
     - 'lib/puppet/parser/functions/getvar.rb'
     - 'lib/puppet/parser/functions/has_interface_with.rb'
 
-# Offense count: 62
-# Configuration parameters: IgnoreSharedExamples.
+# Offense count: 6
+# Configuration parameters: AllowedMethods, AllowedPatterns, CountRepeatedAttributes.
+Metrics/AbcSize:
+  Max: 44
+
+# Offense count: 21
+# Configuration parameters: CountComments, CountAsOne, AllowedMethods, AllowedPatterns, inherit_mode.
+# AllowedMethods: refine
+Metrics/BlockLength:
+  Max: 162
+
+# Offense count: 5
+# Configuration parameters: AllowedMethods, AllowedPatterns.
+Metrics/CyclomaticComplexity:
+  Max: 15
+
+# Offense count: 11
+# Configuration parameters: CountComments, CountAsOne, AllowedMethods, AllowedPatterns.
+Metrics/MethodLength:
+  Max: 37
+
+# Offense count: 1
+# Configuration parameters: CountComments, CountAsOne.
+Metrics/ModuleLength:
+  Max: 104
+
+# Offense count: 4
+# Configuration parameters: AllowedMethods, AllowedPatterns.
+Metrics/PerceivedComplexity:
+  Max: 19
+
+# Offense count: 2
+# Configuration parameters: ForbiddenDelimiters.
+# ForbiddenDelimiters: (?i-mx:(^|\s)(EO[A-Z]{1}|END)(\s|$))
+Naming/HeredocDelimiterNaming:
+  Exclude:
+    - 'lib/puppet/parser/functions/pick.rb'
+    - 'spec/functions/stdlib_deferrable_epp_spec.rb'
+
+# Offense count: 2
+# Configuration parameters: MinNameLength, AllowNamesEndingInNumbers, AllowedNames, ForbiddenNames.
+# AllowedNames: as, at, by, cc, db, id, if, in, io, ip, of, on, os, pp, to
+Naming/MethodParameterName:
+  Exclude:
+    - 'spec/functions/pick_default_spec.rb'
+
+# Offense count: 8
+# This cop supports safe autocorrection (--autocorrect).
+# Configuration parameters: PreferredName.
+Naming/RescuedExceptionsVariableName:
+  Exclude:
+    - 'lib/puppet/functions/parsehocon.rb'
+    - 'lib/puppet/functions/parsepson.rb'
+    - 'lib/puppet/parser/functions/loadjson.rb'
+    - 'lib/puppet/parser/functions/loadyaml.rb'
+    - 'lib/puppet/parser/functions/num2bool.rb'
+    - 'lib/puppet/parser/functions/validate_cmd.rb'
+
+# Offense count: 1
+# Configuration parameters: EnforcedStyle, CheckMethodNames, CheckSymbols, AllowedIdentifiers, AllowedPatterns.
+# SupportedStyles: snake_case, normalcase, non_integer
+# AllowedIdentifiers: capture3, iso8601, rfc1123_date, rfc822, rfc2822, rfc3339, x86_64
+Naming/VariableNumber:
+  Exclude:
+    - 'spec/functions/delete_undef_values_spec.rb'
+
+# Offense count: 1
+# This cop supports unsafe autocorrection (--autocorrect-all).
+Performance/MapCompact:
+  Exclude:
+    - 'lib/puppet/parser/functions/values_at.rb'
```

Comment:
Fixing this one would be more than just 'unsafe'.  This is a ruby 2.7 specific example and Puppet Server 7 is only ruby 2.6 compatible.

Reply:
Yes, there is certain rubocop rules that we have been avoiding during our updates. Mainly those that break the behaviour of the module right away. We will be amending code as soon as possible.

Reply:
That would be interesting, because puppetlabs-concat already uses Ruby 2.7 features: https://github.com/puppetlabs/puppetlabs-concat/blob/7d9fd72f22165b41328fdb29a6a15ea56d275e75/lib/puppet/type/concat_file.rb#L193 uses filter_map.

I'm really wondering about the Ruby 2.6 part. If that's true, then it's going to be very messy.

Reply:
Maybe that bit only ever gets executed on the agent, and in this case, it's still loadable by ruby 2.6 even if execution would fail??

Reply:
Its a very tricky situation as this revelation has come when over 80% of the modules had already undergone the TargetRubyVersion 2.7. Nevertheless, probably only the very first few will modules to be worked on will present problems as we started to reduce the workload each module represented by moving onto a "only deal with autocorrections, ignore problematic cop rules" policy when we realised how limited on time we were.